### DISCUSSION TITLE
Should external dependencies like a CRUD service be part of react query key?

### DISCUSSION THREAD

**Discussion by martynasd123** [TanStack/query#8061] [Q&A]
So I have this crud service interface:

```ts
export interface StuffService {
  getStuff(id: number): Promise<Stuff>;
}

```

I have different implementations of this service depending on whether I'm rendering the component in storybook, tests or production (production implementation uses axios to fetch data, while others just return some static data for testing)

My "fake" service looks like this:

```ts
class MockStuffService implements StuffService {
  private readonly data: Stuff;

  constructor(data: Stuff) {
    super();
    this.data = data;
  }

  async getStuff(id: number): Promise<Stuff> {
    return this.data;
  }
}
```

And this is how I'm using it
```ts

const id = ....

const stuffService = useContext(StuffService)

const { data: stuff } = useQuery(['get-stuff', id], async () => {
  return stuffService.getStuff(id);
});
```

The problem is that when I am wiring up the `MockStuffService` in storybook, I am re-creating it every render. I know I could modify it in-place, but that would be a bit more awkward. So for the fetchFn to get updated, I have to add `stuffService` in the key list:

```ts

const id = ....

const stuffService = useContext(StuffService)

const { data: stuff } = useQuery(['get-stuff', id, stuffService], async () => {
  return stuffService.getStuff(id);
});
```

This doesn't feel right. I know that under the hood react-query uses hashing to check if the function needs to be re-created, and stuffService might be large. Ideally I would want it to be compared by reference. Is there a better, more straightforward way to do this?

**Comment by martynasd123** [comment]
I have prepared a [minimal example](https://codesandbox.io/p/sandbox/react-typescript-forked-hd38tt?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cm164m91f00062e69p2co00h9%2522%252C%2522sizes%2522%253A%255B100%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cm164m91f00022e692qy5w78i%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cm164m91f00032e692xypvx6d%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cm164m91f00052e69azqoj4ff%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cm164m91f00022e692qy5w78i%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm164m91f00012e69bkcictp8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cm164muml001u2e69lw7x2mz4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A40%252C%2522startColumn%2522%253A17%252C%2522endLineNumber%2522%253A40%252C%2522endColumn%2522%253A19%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522cm164m91f00022e692qy5w78i%2522%252C%2522activeTabId%2522%253A%2522cm164muml001u2e69lw7x2mz4%2522%257D%252C%2522cm164m91f00052e69azqoj4ff%2522%253A%257B%2522id%2522%253A%2522cm164m91f00052e69azqoj4ff%2522%252C%2522activeTabId%2522%253A%2522cm164rg8700132e69isikckwq%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cm164m91f00042e6940lfu4g3%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522id%2522%253A%2522cm164me9c00142e69ft7zz7kl%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522id%2522%253A%2522cm164rg8700132e69isikckwq%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522cm164m91f00032e692xypvx6d%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cm164m91f00032e692xypvx6d%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D). I found that this only happens with react query v3, but doesn't with @tanstack/react-query.

The fetched value isn't updated, even though query is invalidated. It still calls the service, but not the new service instance with new "test" parameter. This isn't the case if service is added to the query key array

**Reply by antoantonydh** [reply]
Yes, same issue for me as well, and I am also using v3. In my case, we have Axios client created via context. Because we have micro frontend setup using module federation and each plugin has its own new axios instance. When the plugin unmounts and mounts again, it gets called to the old instance of Axios client and not the new one.
